packages/mui-joy/src/Checkbox/Checkbox.tsx

```diff
@@ -0,0 +1,249 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { OverridableComponent } from '@mui/types';
+import { unstable_capitalize as capitalize } from '@mui/utils';
+import { unstable_composeClasses as composeClasses } from '@mui/base';
+import { useSwitch } from '@mui/base/SwitchUnstyled';
+import { styled, useThemeProps } from '../styles';
+import checkboxClasses, { getCheckboxUtilityClass } from './checkboxClasses';
+import { CheckboxProps, CheckboxTypeMap } from './CheckboxProps';
+import CheckIcon from '../internal/svg-icons/Check';
+import IndeterminateIcon from '../internal/svg-icons/HorizontalRule';
+
+const useUtilityClasses = (ownerState: CheckboxProps & { focusVisible: boolean }) => {
+  const { checked, disabled, focusVisible, color, variant, size } = ownerState;
+
+  const slots = {
+    root: [
+      'root',
+      checked && 'checked',
+      disabled && 'disabled',
+      focusVisible && 'focusVisible',
+      variant && `variant${capitalize(variant)}`,
+      color && `color${capitalize(color)}`,
+      size && `size${capitalize(size)}`,
+    ],
+    input: ['input'],
+  };
+
+  return composeClasses(slots, getCheckboxUtilityClass, {});
+};
+
+const CheckboxRoot = styled('span', {
+  name: 'MuiCheckbox',
+  slot: 'Root',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: CheckboxProps }>(({ theme, ownerState }) => {
+  return [
+    {
+      ...(ownerState.size === 'sm' && {
+        '--Checkbox-size': '1rem',
```

Comment:
I think we can create a theme value for different checkbox sizes and use them instead?

Reply:
Check out https://github.com/mui/material-ui/pull/31273#issuecomment-1057991542